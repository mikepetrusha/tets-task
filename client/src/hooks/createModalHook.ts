import { FC, memo, useCallback, useRef } from 'react';

import { useDialogService } from '@/services/dialog';

export const createModalHook = <T>(Component: (props: T) => FC) => {
  return function useDialog() {
    const idRef = useRef<string>();

    const closeDialog = useCallback(() => {
      if (idRef.current) {
        useDialogService.getState().closeDialog(idRef.current);
      }
    }, []);

    const openDialog = useCallback(
      (props?: Omit<T, 'onClose'>) => {
        idRef.current = useDialogService
          .getState()
          .openDialog(memo(Component({ ...props, onClose: closeDialog } as T)));
      },
      [closeDialog]
    );

    return [openDialog, closeDialog] as const;
  };
};
