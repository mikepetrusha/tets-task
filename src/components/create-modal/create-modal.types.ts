export interface ICreateModalProps {
  title: string;
  onClose: () => void;
}

export interface ICreateModalInput {
  email: string;
  firstname: string;
  lastname: string;
}
