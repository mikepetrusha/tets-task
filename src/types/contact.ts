export interface IContact {
  id: string;
  email: string;
  firstname: string;
  lastname: string;
}
