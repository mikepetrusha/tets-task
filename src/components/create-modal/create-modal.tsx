import { useEffect, useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';

import { createModalHook } from '@/hooks/createModalHook';
import { useContactsService } from '@/services/contacts';

import { Button, Input, Spinner } from '../ui';
import { Modal } from '../ui/modal';
import * as Styled from './create-modal.styles';
import { ICreateModalInput, ICreateModalProps } from './create-modal.types';

export const CreateModal = ({ onClose, title }: ICreateModalProps) => {
  const [isLoading, setIsLoading] = useState(true);
  const { addContact } = useContactsService();

  const handleCreate = (data: ICreateModalInput) => {
    addContact(data);
    onClose();
  };

  const {
    register,
    handleSubmit,
    formState: { isDirty, isValid }
  } = useForm<ICreateModalInput>();
  const onSubmit: SubmitHandler<ICreateModalInput> = (data) => handleCreate(data);

  useEffect(() => {
    setIsLoading(false);
  }, []);

  if (isLoading)
    return (
      <Styled.Overlay>
        <Spinner />
      </Styled.Overlay>
    );

  return (
    <>
      <Modal title={title} onClose={onClose} open>
        <form onSubmit={handleSubmit(onSubmit)}>
          <Styled.FormWrapper>
            <Styled.NameFormWrapper>
              <Input {...register('firstname', { required: true })} placeholder="Vorname*" />
              <Input {...register('lastname', { required: true })} placeholder="Nachname*" />
            </Styled.NameFormWrapper>
            <Input type="email" {...register('email', { required: true })} placeholder="E-Mail*" />
          </Styled.FormWrapper>
          <Styled.ModalActions>
            <Button variant="outlined" onClick={onClose}>
              ABBRECHEN
            </Button>
            <Button disabled={!isDirty || !isValid} type="submit">
              SPEICHERN
            </Button>
          </Styled.ModalActions>
        </form>
      </Modal>
    </>
  );
};

export const useCreateModal = createModalHook<ICreateModalProps>((props) => () => (
  <CreateModal {...props} />
));
