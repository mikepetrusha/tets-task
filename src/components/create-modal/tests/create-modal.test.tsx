import renderer from 'react-test-renderer';

import { CreateModal } from '../create-modal';
import { createModalHook } from '@/hooks/createModalHook';
import { ICreateModalProps } from '../create-modal.types';
import { act } from '@testing-library/react';

it('renders correctly', () => {
  const tree = renderer
    .create(<CreateModal onClose={() => console.log('test')} title="test" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders correctly', async () => {
  await act(() => {
    const useCreateModalTest = createModalHook<ICreateModalProps>((props) => () => (
      <CreateModal {...props} />
    ));

    const TestComp = () => {
      const [openModal] = useCreateModalTest();
      openModal();
      /* fire events that update state */
      return <div>test</div>;
    };
    const tree = renderer.create(<TestComp />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
