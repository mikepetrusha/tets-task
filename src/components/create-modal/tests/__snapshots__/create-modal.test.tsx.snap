// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-csKJxZ hIcYGP"
>
  <div
    className="sc-kFCroH fvwTMG"
    data-testid="test-spinner"
  >
    <div
      className="sc-irLvIq bvdhtb"
    />
  </div>
</div>
`;

exports[`renders correctly 2`] = `null`;
